### ISSUE TITLE
Panzer but think it is really tpetra

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#6781]
## Bug Report
@trilinos/tpetra

If you build panzer with debug-openmp and run the tests you will see a ton fail with an assert

```
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false
Kokkos::DualView::modify_host ERROR: Concurrent modification of host and device views in DualView "MV::DualView"
[s993289:11497] *** Process received signal ***
[s993289:11497] Signal: Aborted (6)
[s993289:11497] Signal code:  (-6)
Kokkos::DualView::modify_host ERROR: Concurrent modification of host and device views in DualView "MV::DualView"
[s993289:11498] *** Process received signal ***
[s993289:11498] Signal: Aborted (6)
[s993289:11498] Signal code:  (-6)

```

Assume this is caused by the checks added by @ndellingwood on the 6th...

### Steps to Reproduce

build develop with gnu-openmp on rhel7 (probably all machines)
run panzer test

(DESKTOP)-dir-gnu-debug-openmp-shared [GNU-7.2.0]$ 58 $ ctest -R PanzerDofMgr_scaling_test

see bug

**Comment by bathmatt** [comment]
So, tpetra tests pass as do the muelu tests.  So, it looks more like a panzer bug that got exposed with the check

**Comment by mhoemmen** [comment]
@bathmatt, @ndellingwood created a PR that might fix this: https://github.com/trilinos/Trilinos/pull/6768 .  It's in develop now.

The cause was that Kokkos' debug settings were not enabling DualView flag checks by default.  They do now.  @ndellingwood noticed the test failures and changed Tpetra to fix them.

**Comment by ndellingwood** [comment]
@bathmatt I can try and help look into this later (I'm out today but had a chance to check email and github and saw this). What Trilinos branch and SHA is this occurring with and what configuration + system?
The error message you posted is occurring because a DualView's  device flags were marked modified by `modify_device`, and then the host was later marked modified by `modify_host` without a synchronization or clearing of flags in between (i.e. "Concurrent modification of host and device views in DualView")

A couple things that can help chase this down:
If it is occurring in Tpetra, setting these environment variables may help hone in on the trouble spot
`export TPETRA_DEBUG=1; export TPETRA_VERBOSE=1` 

To help with debugging #6768 I used the kokkos-tools tool kernel-logger, this helps because it doesn't require rebuilding Trilinos and when the test dies it you'll see the label for the last kernel that was launched. To use, compile the tool and then set the environment variable
`export KOKKOS_PROFILE_LIBRARY=${HOME}/kokkos-tools/src/tools/kernel-logger/kp_kernel_logger.so`
https://github.com/kokkos/kokkos-tools/wiki/KernelLogger

**Comment by bartlettroscoe** [comment]
FYI: As of yesterday, we are seeing 100% passing Panzer tests in all `openmp-debug` builds on all platforms as shown in [this query](https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&date=2020-02-06&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=subprojects&compare2=93&value2=Panzer&field3=block&field3count=2&field3field1=buildname&field3compare1=63&field3value1=openmp-&field3field2=buildname&field3compare2=63&field3value2=openmp_&field4=block&field4count=2&field4field1=buildname&field4compare1=63&field4value1=dbg&field4field2=buildname&field4compare2=63&field4value2=debug).  Could there be a difference with the EMPIRE Trilinos configuration?

**Comment by ndellingwood** [comment]
@bathmatt I think I found the spot in Panzer where this is occurring, prepping a PR now

**Comment by ndellingwood** [comment]
@bathmatt @bartlettroscoe  PR #6782 submitted with a fix

**Comment by bathmatt** [comment]
thanks, i’ll test it out

**Comment by jmgate** [comment]
@bathmatt, how are you configuring/building Trilinos?  We didn't see this failure in our Trilinos sync pipeline, so what are you doing differently?

**Comment by ndellingwood** [comment]
Did PR #6782 resolve the issue?

**Comment by rppawlo** [comment]
Yes. Thanks @ndellingwood !

**Comment by ndellingwood** [comment]
@jmgate I'm assuming the tests that exposed this issue enabled either `Kokkos_ENABLE_DEBUG=ON` or `Kokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=ON`

**Comment by bartlettroscoe** [comment]
Looks like EMPIRE just get an updated version of Trilinos 'develop' as their repo now shows:

```
*   27a0d05 "Merge 'trilinos/Trilinos:develop' (6b4d948) into 'EM-Plasma/Trilinos:develop' (1739355)."
|\  Author: Jenkins Pipeline <jenkins@jenkins-srn.sandia.gov>
| | Date:   Fri Feb 14 08:46:29 2020 -0700 (6 hours ago)
| |     
| *   6b4d948 "Merge Pull Request #6759 from mayrmt/Trilinos/regionMG-speedup-interface-scaling"
| |\  Author: trilinos-autotester <trilinos@sandia.gov>
| | | Date:   Thu Feb 13 14:21:16 2020 -0700 (24 hours ago)
| | |    
| | * 18079fc "MueLu: remove obsolete zero-out of vector in regMG"
| | | Author: Matthias Mayr <matthias.mayr@unibw.de>
| | | Date:   Mon Feb 10 17:43:44 2020 +0100 (4 days ago)
| | | 
| | | M packages/muelu/research/mmayr/composite_to_regions/src/SetupRegionSmoothers_def.hpp
```

So this is good to close!

Now we just need to get SPARC to fix their code to absorb the Kokkos 2.99 changes and then we can get back to to a good place.
